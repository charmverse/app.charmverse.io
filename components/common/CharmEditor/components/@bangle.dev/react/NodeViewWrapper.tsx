import type { NodeView, NodeViewProps } from '@bangle.dev/core';
import React from 'react';

export type RenderNodeViewsFunction = (
  props: NodeViewProps & { children: React.ReactNode },
) => React.ReactNode;

interface PropsType {
  nodeView: NodeView;
  renderNodeViews: RenderNodeViewsFunction;
  nodeViewUpdateStore: WeakMap<NodeView, () => void>;
}

interface StateType {
  nodeViewProps: NodeViewProps;
}

export class NodeViewWrapper extends React.Component<PropsType, StateType> {
  update: () => void;

  attachToContentDOM: (reactElement: HTMLDivElement) => void;

  constructor (props: PropsType) {
    super(props);
    this.update = () => {
      this.setState(() => ({
        nodeViewProps: props.nodeView.getNodeViewProps()
      }));
    };

    this.attachToContentDOM = (reactElement) => {
      if (!reactElement) {
        return;
      }
      const contentDOM = this.props.nodeView.contentDOM!;
      // Since we do not control how many times this callback is called
      // make sure it is not already mounted.
      if (!reactElement.contains(contentDOM)) {
        // If contentDOM happens to be mounted to someone else
        // remove it from there.
        if (contentDOM.parentNode) {
          contentDOM.parentNode.removeChild(contentDOM);
        }
        reactElement.appendChild(contentDOM);
      }
    };
    // So that we can directly update the nodeView without the mess
    // of prop forwarding.
    // What about updating the wrong nodeView ?
    // It is okay because a nodeView and this ReactComponent will always
    // have a 1:1 mapping. This is guaranteed because you use `nodeView` instance
    // to generate a react key. See the usage of this component in ./ReactEditor.js
    props.nodeViewUpdateStore.set(props.nodeView, this.update);
    this.state = { nodeViewProps: this.props.nodeView.getNodeViewProps() };
  }

  getChildren () {
    if (!this.props.nodeView.contentDOM) {
      return null;
    }

    // if (this.state.nodeViewProps.node.isInline) {
    //   return (
    //     // The bangle-nv-content is needed to keep the content take space
    //     // or else browsers will collapse it, making it hard to type
    //     <span
    //       className="bangle-nv-child-container"
    //       ref={this.attachToContentDOM}
    //     />
    //   );
    // }

    return (
      <div
        className='bangle-nv-child-container'
        ref={this.attachToContentDOM}
      />
    );
  }

  render () {
    const element = this.props.renderNodeViews({
      ...this.state.nodeViewProps,
      children: this.getChildren()
    });
    if (!element) {
      throw new Error(
        `renderNodeView must handle rendering for node of type "${this.state.nodeViewProps.node.type.name}"`
      );
    }
    return element;
  }
}
