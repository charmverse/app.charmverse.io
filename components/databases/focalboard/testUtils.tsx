// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {IntlProvider} from 'react-intl'
import React from 'react'
import {DndProvider} from 'react-dnd'
import {HTML5Backend} from 'react-dnd-html5-backend'
import configureStore, {MockStoreEnhanced} from 'redux-mock-store'
import {Middleware} from 'redux'

import {Block} from './blocks/block'

export const wrapIntl = (children?: React.ReactNode): JSX.Element => <IntlProvider locale='en'>{children}</IntlProvider>
export const wrapDNDIntl = (children?: React.ReactNode): JSX.Element => {
    return (
        <DndProvider backend={HTML5Backend}>
            {wrapIntl(children)}
        </DndProvider>
    )
}

export function mockDOM(): void {
    window.focus = jest.fn()
    document.createRange = () => {
        const range = new Range()
        range.getBoundingClientRect = jest.fn()
        range.getClientRects = () => {
            return {
                item: () => null,
                length: 0,
                [Symbol.iterator]: jest.fn(),
            }
        }
        return range
    }
}
export function mockMatchMedia(result: {matches: boolean}): void {
    // We check if system preference is dark or light theme.
    // This is required to provide it's definition since
    // window.matchMedia doesn't exist in Jest.
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation(() => {
            return result

            // return ({
            //     matches: true,
            // })
        }),
    })
}

export function mockStateStore(middleware:Middleware[], state:unknown): MockStoreEnhanced<unknown, unknown> {
    const mockStore = configureStore(middleware)
    return mockStore(state)
}

export type BlocksById<BlockType> = {[key: string]: BlockType}

export function blocksById<BlockType extends Block>(blocks: Array<BlockType>): BlocksById<BlockType> {
    return blocks.reduce((res, block) => {
        res[block.id] = block
        return res
    }, {} as BlocksById<BlockType>)
}
