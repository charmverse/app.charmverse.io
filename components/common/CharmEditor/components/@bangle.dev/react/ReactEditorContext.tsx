import type { EditorView } from 'prosemirror-view';
import type { Dispatch, ReactNode, SetStateAction } from 'react';
import { createContext, useContext, useMemo, useState } from 'react';

// keep track of the focused page (may be different from what's in the URL or header)
type ICharmEditorView = {
  view: EditorView | null;
  setView: (view: EditorView) => void;
};

const CharmEditorView = createContext<Readonly<ICharmEditorView>>({
  view: null,
  setView: () => {}
});

export function CharmEditorViewProvider({ children }: { children: ReactNode }) {
  const [view, setView] = useState<EditorView | null>(null);

  const value: ICharmEditorView = useMemo(
    () => ({
      view,
      setView
    }),
    [view, setView]
  );

  return <CharmEditorView.Provider value={value}>{children}</CharmEditorView.Provider>;
}

export const useCharmEditorView = () => useContext(CharmEditorView);
