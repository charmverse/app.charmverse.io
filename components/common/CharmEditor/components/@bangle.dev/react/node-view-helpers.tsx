import { objectUid } from '@bangle.dev/utils';
import { log } from '@charmverse/core/log';
import { useEffect, useState } from 'react';
import type { RefObject } from 'react';
import { flushSync } from 'react-dom';

import type { NodeView, RenderHandlers } from '../core/node-view';
import { saveRenderHandlers } from '../core/node-view';

export const nodeViewUpdateStore = new WeakMap();

type NodeViewsUpdater = (nodeViewUpdateStore: NodeView[]) => NodeView[];
type UpdateNodeViewsFunction = (updater: NodeViewsUpdater) => void;

export const nodeViewRenderHandlers = (updateNodeViews: UpdateNodeViewsFunction): RenderHandlers => ({
  create: (nodeView, _nodeViewProps) => {
    log.debug('create node', objectUid.get(nodeView)); // , new Error().stack);
    updateNodeViews((nodeViews) => [...nodeViews, nodeView]);
  },
  update: (nodeView, _nodeViewProps) => {
    log.debug('update node', objectUid.get(nodeView));
    const updateCallback = nodeViewUpdateStore.get(nodeView);
    // If updateCallback is undefined (which can happen if react took long to mount),
    // we are still okay, as the latest nodeViewProps will be accessed whenever it mounts.
    if (updateCallback) {
      updateCallback();
    }
  },
  destroy: (nodeView) => {
    // log.debug('destroy', objectUid.get(nodeView));
    updateNodeViews((nodeViews) => nodeViews.filter((n) => n !== nodeView));
  }
});

export function useNodeViews(ref: RefObject<HTMLElement>) {
  const [nodeViews, setNodeViews] = useState<NodeView[]>([]);
  useEffect(() => {
    // save the renderHandlers in the dom to decouple nodeView instantiating code
    // from the editor. Since PM passing view when nodeView is created, the author
    // of the component can get the handler reference from `getRenderHandlers(view)`.
    // Note: this assumes that the pm's dom is the direct child of `editorRenderTarget`.
    let destroyed = false;
    saveRenderHandlers(
      ref.current!,
      nodeViewRenderHandlers((cb) => {
        if (!destroyed) {
          // make sure that updates run sequentially for prosemiror, or we get infinite recursion of node views being created and destroyed.
          // See also: https://github.com/ueberdosis/tiptap/pull/2985
          flushSync(() => {
            // use callback variant of setState to
            // always get freshest nodeViews.
            setNodeViews((_nodeViews) => cb(_nodeViews));
          });
        }
      })
    );
    return () => {
      destroyed = true;
    };
  }, [ref]);

  return nodeViews;
}
