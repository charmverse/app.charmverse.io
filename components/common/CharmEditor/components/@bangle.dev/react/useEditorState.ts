import { useState } from 'react';

import type { BangleEditorStateProps } from '../core/bangle-editor-state';
import { BangleEditorState } from '../core/bangle-editor-state';

export function useEditorState(props: BangleEditorStateProps) {
  if (props.plugins && typeof props.plugins !== 'function') {
    throw new Error('plugins error: plugins must be a function');
  }
  const [state] = useState(
    () =>
      // Instantiate the editorState once and keep using that instance
      // on subsequent renders.
      // Passing a callback in useState lazy calls the
      // functions on the first render and never again.
      new BangleEditorState(props)
  );
  return state;
}
