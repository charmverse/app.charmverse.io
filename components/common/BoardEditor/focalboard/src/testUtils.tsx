import type { PageMeta } from '@charmverse/core/pages';
import React from 'react';
import { DndProvider } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';
import { IntlProvider } from 'react-intl';
import type { Middleware } from 'redux';
import type { MockStoreEnhanced } from 'redux-mock-store';
import configureStore from 'redux-mock-store';
import { v4 } from 'uuid';

import { PagesContext } from 'hooks/usePages';
import type { Block } from 'lib/focalboard/block';
import { pageStubToCreate } from 'testing/generatePageStub';
import { AppThemeProvider } from 'theme/AppThemeProvider';

export const wrapIntl = (children?: React.ReactNode): JSX.Element => (
  <AppThemeProvider>
    <IntlProvider locale='en'>{children}</IntlProvider>
  </AppThemeProvider>
);
export const wrapDNDIntl = (children?: React.ReactNode): JSX.Element => {
  return <DndProvider backend={HTML5Backend}>{wrapIntl(children)}</DndProvider>;
};

export const wrapPagesProvider = (cardId: string | string[], children: React.ReactNode) => {
  const pages: Record<string, PageMeta> = {};

  const cardIds = Array.isArray(cardId) ? cardId : [cardId];

  cardIds.forEach((_cardId) => {
    pages[_cardId] = pageStubToCreate({ spaceId: v4(), createdBy: v4(), id: _cardId, path: _cardId }) as PageMeta;
  });

  return (
    <PagesContext.Provider
      value={
        {
          pages
        } as any
      }
    >
      {children}
    </PagesContext.Provider>
  );
};

export function mockDOM(): void {
  window.focus = jest.fn();
  document.createRange = () => {
    const range = new Range();
    range.getBoundingClientRect = jest.fn();
    range.getClientRects = () => {
      return {
        item: () => null,
        length: 0,
        [Symbol.iterator]: jest.fn()
      };
    };
    return range;
  };
}
export function mockMatchMedia(result: { matches: boolean }): void {
  // We check if system preference is dark or light theme.
  // This is required to provide it's definition since
  // window.matchMedia doesn't exist in Jest.
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(() => {
      return result;

      // return ({
      //     matches: true,
      // })
    })
  });
}

export function mockStateStore(middleware: Middleware[], state: unknown): MockStoreEnhanced<unknown, unknown> {
  const mockStore = configureStore(middleware);
  return mockStore(state);
}

export type BlocksById<BlockType> = { [key: string]: BlockType };

export function blocksById<BlockType extends Block>(blocks: BlockType[]): BlocksById<BlockType> {
  return blocks.reduce((res, block) => {
    res[block.id] = block;
    return res;
  }, {} as BlocksById<BlockType>);
}
