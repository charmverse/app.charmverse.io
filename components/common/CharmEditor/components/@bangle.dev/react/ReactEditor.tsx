import type {
  BangleEditorProps as CoreBangleEditorProps
} from '@bangle.dev/core';
import {
  BangleEditor as CoreBangleEditor
} from '@bangle.dev/core';
import type { Plugin, Transaction } from '@bangle.dev/pm';
import { EditorViewContext } from '@bangle.dev/react';
import { nodeViewUpdateStore, useNodeViews } from '@bangle.dev/react/node-view-helpers';
import { objectUid } from '@bangle.dev/utils';
import type { ReactNode, RefObject } from 'react';
import React, { useEffect, useImperativeHandle, useRef, useState } from 'react';
import reactDOM from 'react-dom';

import { useUser } from 'hooks/useUser';
import log from 'lib/log';
import { isTouchScreen } from 'lib/utilities/browser';

import { amendTransaction } from '../../suggestions/track/amendTransaction';

import { NodeViewWrapper } from './NodeViewWrapper';
import type { RenderNodeViewsFunction } from './NodeViewWrapper';

interface BangleEditorProps<PluginMetadata = any>
  extends CoreBangleEditorProps<PluginMetadata> {
  id?: string;
  children?: React.ReactNode;
  renderNodeViews?: RenderNodeViewsFunction;
  className?: string;
  style?: React.CSSProperties;
  onReady?: (editor: CoreBangleEditor<PluginMetadata>) => void;
  editorRef?: RefObject<HTMLDivElement>;
  // Components that should be placed underneath the editor
  placeholderComponent?: ReactNode;
  enableSuggestions?: boolean; // requires trackChanges to be true
  trackChanges?: boolean;
}

const updatePluginWatcher = (editor: CoreBangleEditor) => {
  return (watcher: Plugin, remove = false) => {
    if (editor.destroyed) {
      return;
    }

    let state = editor.view.state;

    const newPlugins = remove
      ? state.plugins.filter((p) => p !== watcher)
      : [...state.plugins, watcher];

    state = state.reconfigure({
      plugins: newPlugins
    });

    editor.view.updateState(state);
  };
};

export const BangleEditor = React.forwardRef<
  CoreBangleEditor | undefined,
  BangleEditorProps
>(
  (
    {
      id,
      state,
      children,
      focusOnInit = !isTouchScreen(),
      pmViewOpts,
      renderNodeViews,
      className,
      style,
      onReady = () => {},
      placeholderComponent,
      editorRef,
      enableSuggestions = false,
      trackChanges = false
    },
    ref
  ) => {
    const renderRef = useRef<HTMLDivElement>(null);
    const onReadyRef = useRef(onReady);
    const editorViewPayloadRef = useRef({
      state,
      focusOnInit,
      pmViewOpts,
      enableSuggestions
    });
    const [editor, setEditor] = useState<CoreBangleEditor>();
    const nodeViews = useNodeViews(renderRef);
    const { user } = useUser();

    if (enableSuggestions && !trackChanges) {
      log.error('CharmEditor: Suggestions require trackChanges to be enabled');
    }

    // set current
    editorViewPayloadRef.current.enableSuggestions = enableSuggestions;

    useImperativeHandle(
      ref,
      () => {
        return editor;
      },
      [editor]
    );

    useEffect(() => {
      const _editor = new CoreBangleEditor(
        renderRef.current!,
        editorViewPayloadRef.current
      );
      (_editor.view as any)._updatePluginWatcher = updatePluginWatcher(_editor);
      if (trackChanges) {
        (_editor.view as any)._props.dispatchTransaction = (transaction: Transaction) => {
          const view = _editor.view;
          const trackedUser = { id: user?.id ?? '', username: user?.username ?? '' };
          const trackedTr = amendTransaction(transaction, view.state, trackedUser, editorViewPayloadRef.current.enableSuggestions);
          const { state: newState } = view.state.applyTransaction(trackedTr);
          view.updateState(newState);
        };
      }
      onReadyRef.current(_editor);
      setEditor(_editor);
      return () => {
        _editor.destroy();
      };
    }, [ref]);

    if (nodeViews.length > 0 && renderNodeViews == null) {
      throw new Error(
        'When using nodeViews, you must provide renderNodeViews callback'
      );
    }

    return (
      <EditorViewContext.Provider value={editor?.view as any}>
        <div ref={editorRef} className='bangle-editor-core'>
          {editor ? children : null}
          <div ref={renderRef} id={id} className={className} style={style} />
          {editor ? placeholderComponent : null}
        </div>
        {nodeViews.map((nodeView) => {
          return reactDOM.createPortal(
            <NodeViewWrapper
              nodeViewUpdateStore={nodeViewUpdateStore}
              nodeView={nodeView}
              renderNodeViews={renderNodeViews!}
            />,
              nodeView.containerDOM!,
              objectUid.get(nodeView)
          );
        })}
      </EditorViewContext.Provider>
    );
  }
);
