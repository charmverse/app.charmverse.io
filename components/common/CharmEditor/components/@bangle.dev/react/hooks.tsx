import type { EditorView } from '@bangle.dev/pm';
import { Plugin, PluginKey } from '@bangle.dev/pm';
import { rafSchedule } from '@bangle.dev/utils';
import { useContext, useEffect, useState } from 'react';

import type { BangleEditorStateProps } from 'components/common/CharmEditor/components/@bangle.dev/core/bangle-editor-state';

import { BangleEditorState } from '../core/bangle-editor-state';

import { EditorViewContext } from './ReactEditor';

export function useEditorState(props: BangleEditorStateProps) {
  if (props.plugins && typeof props.plugins !== 'function') {
    throw new Error('plugins error: plugins must be a function');
  }

  const [state] = useState(
    () =>
      // Instantiate the editorState once and keep using that instance
      // on subsequent renders.
      // Passing a callback in useState lazy calls the
      // functions on the first render and never again.
      new BangleEditorState(props)
  );

  return state;
}

export function usePluginState(pluginKey: PluginKey, throttle = false) {
  const view = useEditorViewContext();
  const [state, setState] = useState(pluginKey.getState(view.state));

  useEffect(() => {
    let _setState = setState;
    if (throttle) {
      _setState = rafSchedule(setState);
    }
    const plugin = watcherPlugin(pluginKey, _setState);
    (view as any)._updatePluginWatcher(plugin);
    return () => {
      if (throttle) {
        (_setState as ReturnType<typeof rafSchedule>).cancel();
      }
      (view as any)._updatePluginWatcher(plugin, true);
    };
  }, [view, pluginKey, throttle]);

  return state;
}

export function useEditorViewContext(): EditorView {
  const context = useContext(EditorViewContext);
  if (!context) {
    throw new Error('useEditorViewContext must be used within a EditorViewContext');
  }
  return context;
}

function watcherPlugin(pluginKey: PluginKey, setState: (state: any) => void) {
  return new Plugin({
    key: new PluginKey(`withPluginState_${(pluginKey as any).key}`),
    view() {
      return {
        update(view, prevState) {
          const { state } = view;
          if (prevState === state) {
            return;
          }
          const newPluginState = pluginKey.getState(state);

          if (newPluginState !== pluginKey.getState(prevState)) {
            setState(newPluginState);
          }
        }
      };
    }
  });
}
