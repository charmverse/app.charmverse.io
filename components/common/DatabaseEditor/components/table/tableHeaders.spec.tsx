import '@testing-library/jest-dom';
import { render } from '@testing-library/react';

import { TestBlockFactory } from '../../test/testBlockFactory';
import { wrapDNDIntl } from '../../testUtils';

import TableHeaders from './tableHeaders';

jest.mock('@ethereum-attestation-service/eas-sdk', () => ({}));

describe('components/table/TableHeaders', () => {
  const board = TestBlockFactory.createBoard();
  const card = TestBlockFactory.createCard(board);
  const view = TestBlockFactory.createBoardView(board);

  test('should match snapshot', async () => {
    const component = wrapDNDIntl(
      <TableHeaders
        board={board}
        cards={[card]}
        activeView={view}
        views={[view]}
        readOnly={false}
        resizingColumn=''
        offset={0}
        columnRefs={new Map()}
      />
    );

    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });
});
