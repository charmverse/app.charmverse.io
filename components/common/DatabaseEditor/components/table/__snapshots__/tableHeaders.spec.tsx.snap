// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/table/TableHeaders should match snapshot 1`] = `
<div>
  <span
    class="variable variable"
  >
    <div
      class="octo-table-header TableHeaders"
      id="mainBoardHeader"
    >
      <div
        class="octo-table-cell header-cell disable-drag-selection"
        data-test="table-property-text"
        draggable="true"
        style="overflow: unset; width: 100px; opacity: 1; transition: background-color 150ms ease-in-out;"
      >
        <div
          class="MuiStack-root css-19zgtxi-MuiStack-root"
        >
          <div>
            <div>
              <span
                class="Label empty "
              >
                <div
                  style="margin-right: 4px; display: flex;"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-fatq9v-MuiSvgIcon-root"
                    data-testid="SubjectIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M14 17H4v2h10zm6-8H4v2h16zM4 15h16v-2H4zM4 5v2h16V5z"
                    />
                  </svg>
                </div>
                <span
                  aria-label="Title"
                  class="MuiTypography-root MuiTypography-subtitle1 css-1to5392-MuiTypography-root"
                  data-mui-internal-clone-element="true"
                >
                  Title
                </span>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall css-10xxybu-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                />
              </span>
            </div>
          </div>
        </div>
        <div
          class="octo-spacer"
        />
        <div
          class="HorizontalGrip"
          draggable="true"
        />
      </div>
      <div
        class="octo-table-cell header-cell disable-drag-selection"
        style="flex-grow: 1;"
      >
        <button
          class="Button size--small"
          type="button"
        >
          <span>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-qolhez-MuiSvgIcon-root"
              data-test="add-table-prop"
              data-testid="AddIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </span>
</div>
`;
